include/jemalloc/internal/thread_event.h

```diff
@@ -218,6 +218,13 @@ te_ctx_get(tsd_t *tsd, te_ctx_t *ctx, bool is_alloc) {
 	}
 }
 
+JEMALLOC_ALWAYS_INLINE bool
+te_prof_sample_event_lookahead(tsd_t *tsd, size_t usize) {
```

Comment:
I think with a little bit of refactoring, we can get rid of this function: when the thread event module calls prof_alloc_prep, in addition to usize, it can also pass `tsd_thread_allocated_get(tsd) - tsd_thread_allocated_last_event_get(tsd)` (say, in a parameter named `bytes_allocated_since_last_event` or something).

That lets us get rid of the loop where thread event calls prof calls thread event; we sort of decouple how tightly tied together they are. (I also think the resulting condition is a little cleaner; `if (usize + bytes_since_last_event >= tsd_prof_sample_event_wait_get(tsd))` reads a little bit more obviously to me as to what the intent is, although I guess that's debatable).

Reply:
I think there are two things here:

First, there's only `prof_alloc_prep()` -> te look-ahead, rather than te -> `prof_alloc_prep()` -> te. The way of my current design is: te only provides a single write API, the commit, which should always be called at the end and with the right amount of increment; before that, there is some handy event-specific read-only look-ahead API, if the caller is in need (not if te is in need).

Second, in case you're talking about the logic loop, not actual function-call loop, then you're right that te depends on prof and prof also depends on te (the look-ahead), on the module level. But I don't see an easy way to chop either dependency. The te -> prof dependency is by design how te works. The prof -> te look-ahead dependency is also how the te look-ahead works. It's not obvious to me why passing down `bytes_since_last_event` could cut the dependency; the caller just ends up trying to get this `bytes_since_last_event` every time before invoking `prof_alloc_prep()`, which feels in need of a code consolidation.

Thoughts?

Reply:
It's the logic loop that's the concern. The ideal end stat of module dependencies is a DAG.

Stuff in TSD is effectively global.
```
data = f();
g(data);
```
has some amount of extra code relative to:
```
f();
g(); // g and f are secretly communicating through global data
```
But it's sort of a "good" redundancy in that it's really just making explicit what would otherwise be a hidden communication channel. If the computation code is the issue, we can always pull out the `bytes_since_last_event` computation into a helper function.

Reply:
Or put more succinctly: functional purity is a really useful property for testing and reasoning about our internals. We shouldn't give it up without a really big benefit.

Reply:
Makes sense. Made the revision as a separate commit, since it's kind of a refactoring of the refactoring.

Rather than using `bytes_since_last_event` to communicate, I'm still encapsulating the look-ahead inside of the thread event module, and let the caller directly get a boolean `sample_event` and then use it to communicate with prof. I think separating out the `tsd_thread_allocated_get(tsd) - tsd_thread_allocated_last_event_get(tsd)` part and the `if (usize + bytes_since_last_event >= tsd_prof_sample_event_wait_get(tsd))` part is unnecessary; they are thread event specific computations and would better be kept inside of the thread event module.

After doing the DAG conversion, I actually figured that the outcome looks more desired despite the extra function call, due to the following:
- I realized that `prof_alloc_prep()`'s only dependency on `usize` is for the look-ahead. So rather than passing `usize` all the way in, I'm now only passing the `sample_event` flag.
- The caller is now more explicit in obtaining the `sample_event` flag, in that whenever there's a possible `usize` correction, the caller explicitly invokes the look-ahead again (and pass the result to `prof_realloc()`).
- I realized that I'd need to relocate the look-ahead call to the caller anyway in #1777 - at that time I'd need the `surplus` value, and the most convenient way to obtain it is through the look-ahead; if the look-ahead call is buried deep down in `prof_alloc_prep()`, then `surplus` needs to be surfaced all the way back so that `prof_malloc()` / `prof_realloc()` can use it. This is a good example of the practical benefit of the DAG design.